src/docs/sphinx/expr_templates.rst

```diff
@@ -0,0 +1,78 @@
+.. ## Copyright (c) 2019-2020, Lawrence Livermore National Security, LLC and
+.. ## other Serac Project Developers. See the top-level COPYRIGHT file for details.
+.. ##
+.. ## SPDX-License-Identifier: (BSD-3-Clause)
+
+====================
+Expression Templates
+====================
+
+Expression templates are a C++ technique that leverages static polymorphism (with the `Curiously 
```

Comment:
Add a note here to say to only use this for prototyping

Reply:
I don't believe it is only for prototyping.

Reply:
I think with the current [performance hit](https://github.com/LLNL/serac/pull/165#issuecomment-685968525), it's best to keep it for prototyping purposes (at least in hot loops) for now. Once we can beat down these slowdowns, we can put them in production kennels.

Reply:
I view this current functionality similar to using auto differentiation for calculating residual gradients. It's good generating baselines for performant versions.

Reply:
When I spoke to Josh earlier, he said the 20x slowdown was from comparing times from debug expression templates w/ release mfem, which isn't really an honest comparison.

On my home machine, comparing release serac w/ release mfem:
```
[==========] Running 12 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 12 tests from expr_templates
[ RUN      ] expr_templates.basic_add
[       OK ] expr_templates.basic_add (0 ms)
[ RUN      ] expr_templates.basic_add_lambda
[       OK ] expr_templates.basic_add_lambda (0 ms)
[ RUN      ] expr_templates.move_from_temp_lambda
[       OK ] expr_templates.move_from_temp_lambda (0 ms)
[ RUN      ] expr_templates.move_from_temp_vec_lambda
[       OK ] expr_templates.move_from_temp_vec_lambda (0 ms)
[ RUN      ] expr_templates.small_matvec
[       OK ] expr_templates.small_matvec (0 ms)
[ RUN      ] expr_templates.small_mixed_expr
Expression templates took 1.58392 times as long as the raw MFEM calls
[       OK ] expr_templates.small_mixed_expr (9 ms)
[ RUN      ] expr_templates.large_mixed_expr
Expression templates took 1.02729 times as long as the raw MFEM calls
[       OK ] expr_templates.large_mixed_expr (1980 ms)
[ RUN      ] expr_templates.complex_expr_lambda
[       OK ] expr_templates.complex_expr_lambda (0 ms)
[ RUN      ] expr_templates.small_vec_single_add_time_check
Expression templates took 0.916667 times as long as the raw MFEM calls
[       OK ] expr_templates.small_vec_single_add_time_check (0 ms)
[ RUN      ] expr_templates.large_vec_single_add_time_check
Expression templates took 0.84289 times as long as the raw MFEM calls
[       OK ] expr_templates.large_vec_single_add_time_check (23 ms)
[ RUN      ] expr_templates.small_vec_multi_add_time_check
Expression templates took 0.83571 times as long as the raw MFEM calls
[       OK ] expr_templates.small_vec_multi_add_time_check (2 ms)
[ RUN      ] expr_templates.large_vec_multi_add_time_check
Expression templates took 0.309412 times as long as the raw MFEM calls
[       OK ] expr_templates.large_vec_multi_add_time_check (79 ms)
[----------] 12 tests from expr_templates (2094 ms total)
[----------] Global test environment tear-down
[==========] 12 tests from 1 test suite ran. (2094 ms total)
[  PASSED  ] 12 tests.
```

These tests show significant speedups, despite the fact that our expression template implementation is inherently flawed in the sense that we can't modify mfem::Vector to eliminate some unnecessary memory allocations. The exceptional 1.5x slowdown case is from a test with tiny vectors, where the extra overhead from dynamically allocating memory in the loop hurts.

Reply:
Ahhhh that's much better!! I think with these numbers using this in production is not only allowed, it should be encouraged.

Reply:
(that's what I get for trying to understand a technical discussion while chasing my kid!)

Reply:
Also, if we can show how nice this is I can see making the argument to upstream the fix as well as this implementation to mfem.

Reply:
Also, if we can show how nice this is I can see making the argument to upstream the fix as well as this implementation to mfem.

Reply:
For the sake of completeness, comparing debug serac against debug mfem:

```
[==========] Running 12 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 12 tests from expr_templates
[ RUN      ] expr_templates.basic_add
[       OK ] expr_templates.basic_add (0 ms)
[ RUN      ] expr_templates.basic_add_lambda
[       OK ] expr_templates.basic_add_lambda (0 ms)
[ RUN      ] expr_templates.move_from_temp_lambda
[       OK ] expr_templates.move_from_temp_lambda (0 ms)
[ RUN      ] expr_templates.move_from_temp_vec_lambda
[       OK ] expr_templates.move_from_temp_vec_lambda (0 ms)
[ RUN      ] expr_templates.small_matvec
[       OK ] expr_templates.small_matvec (0 ms)
[ RUN      ] expr_templates.small_mixed_expr
Expression templates took 1.17761 times as long as the raw MFEM calls
[       OK ] expr_templates.small_mixed_expr (59 ms)
[ RUN      ] expr_templates.large_mixed_expr
Expression templates took 0.995274 times as long as the raw MFEM calls
[       OK ] expr_templates.large_mixed_expr (6551 ms)
[ RUN      ] expr_templates.complex_expr_lambda
[       OK ] expr_templates.complex_expr_lambda (0 ms)
[ RUN      ] expr_templates.small_vec_single_add_time_check
Expression templates took 2.75278 times as long as the raw MFEM calls
[       OK ] expr_templates.small_vec_single_add_time_check (6 ms)
[ RUN      ] expr_templates.large_vec_single_add_time_check
Expression templates took 6.58961 times as long as the raw MFEM calls
[       OK ] expr_templates.large_vec_single_add_time_check (321 ms)
[ RUN      ] expr_templates.small_vec_multi_add_time_check
Expression templates took 1.61061 times as long as the raw MFEM calls
[       OK ] expr_templates.small_vec_multi_add_time_check (23 ms)
[ RUN      ] expr_templates.large_vec_multi_add_time_check
Expression templates took 2.97274 times as long as the raw MFEM calls
[       OK ] expr_templates.large_vec_multi_add_time_check (1425 ms)
[----------] 12 tests from expr_templates (8386 ms total)
[----------] Global test environment tear-down
[==========] 12 tests from 1 test suite ran. (8386 ms total)
[  PASSED  ] 12 tests.
```

It is true that expression templates tend to not perform as well in debug builds, as they rely pretty heavily on inlining (which is typically disabled in debug builds).

Reply:
Any idea how much of the runtime this part of the code is?  Those numbers are a lot more promising for sure. Especially if they would improve with some changes on the MFEM side.

Reply:
It depends on the physics operator, but these calls would probably be small compared to the global linear solves.